contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,417 @@
+A new `CSIVolumeSource` object will be added to the Kubernetes API. It will be part of the existing `VolumeSource` and `PersistentVolumeSource` objects. This will enable Kubernetes users to use the new volume just like existing volume via a `PersistentVolume` or as a direct reference in a pod.
+
+```GO
+type CSIVolumeSource struct {
+  // Driver is the name of the driver to use for this volume.
+  // Required.
+  Driver string `json:"driver" protobuf:"bytes,1,opt,name=driver"`
+
+  // VolumeHandle is the unique volume name returned by the CSI volume
+  // plugin’s CreateVolume to refer to the volume on all subsequent calls.
+  VolumeHandle string `json:"volumeHandle" protobuf:"bytes,2,opt,name=volumeHandle"`
+
+  // Optional: MountSecretRef is a reference to the secret object containing
+  // sensitive information to pass to the CSI driver during NodePublish. 
+  // This may be empty if no secret is required. If the secret object contains
+  // more than one secret, all secrets are passed.
+  // This is a local object reference so only secrets within the same
+  // namespace as the pod or PVC referencing this volume can be referenced.
+  // +optional
+  MountSecretRef *LocalObjectReference `json:"secretRef,omitempty" protobuf:"bytes,3,opt,name=secretRef"`
+
+  // Optional: AttachSecretRef is a reference to the secret object containing
+  // sensitive information to pass to the CSI driver during ControllerPublish. 
+  // This may be empty if no secret is required. If the secret object contains
+  // more than one secret, all secrets are passed.
+  // +optional
+  AttachSecretRef *ObjectReference `json:"secretRef,omitempty" protobuf:"bytes,3,opt,name=secretRef"`
+
+  // Optional: The value to pass to ControllerPublishVolumeRequest.
+  // Defaults to false (read/write).
+  // +optional
+  ReadOnly bool `json:"readOnly,omitempty" protobuf:"varint,4,opt,name=readOnly"`
+}
+```
+
+#### Internal Interfaces
+
+The in-tree CSI volume plugin will implement the following internal Kubernetes volume interfaces:
+
+1. `VolumePlugin`
+    * Mounting/Unmounting of a volume to a specific path.
+2. `AttachableVolumePlugin`
+    * Attach/detach of a volume to a given node.
+
+Notably, `ProvisionableVolumePlugin` and `DeletableVolumePlugin` are not implemented because provisioning and deleting for CSI volumes is handled by an external provisioner. 
+
+#### Mount and Unmount
+
+The in-tree volume plugin’s mount and unmount methods will trigger the `NodePublishVolume` and `NodeUnpublishVolume` CSI calls via Unix Domain Socket. Kubernetes will generate a unique `target_path` (unique per pod per volume) to pass via `NodePublishVolume` for the CSI plugin to mount the volume. Upon successful completion of the `NodeUnpublishVolume` call (once volume unmount has been verified), Kubernetes will delete the directory.
+
+The Kubernetes volume sub-system does not currently support block volumes (only file), so for alpha, the Kubernetes CSI volume plugin will only support file.
```

Comment:
I am a little confused with the definitions between "raw block volume" and "block volume" in Kubernetes.  So does it mean that in the alpha version Kubernetes will not support the block volume drivers like EBS and Ceph RBD?

Reply:
I agree it's confusing, but in Kubernetes context "raw block volume" and "block volume" usually means the same - access to raw device in pods - and we won't support it in Alpha implementation of CSI plugin.

On the other hand, we will support filesystems on block devices as we do now, i.e. one can implement a CSI volume driver that attaches a block device (say EBS) and mounts it and a pod can use it.

Reply:
@jsafrane Thank your reply. It is very useful for me.